enzyme/Enzyme/Utils.cpp

```diff
@@ -668,6 +668,112 @@ void callMemcpyStridedLapack(llvm::IRBuilder<> &B, llvm::Module &M,
   B.CreateCall(fn, args, bundles);
 }
 
+llvm::CallInst *getorInsertInnerProd(
+    llvm::IRBuilder<> &B, llvm::Module &M, BlasInfo blas, IntegerType *IT,
+    Type *BlasPT, Type *BlasIT, Type *fpTy, llvm::ArrayRef<llvm::Value *> args,
+    const llvm::ArrayRef<llvm::OperandBundleDef> bundles, bool byRef) {
+  assert(fpTy->isFloatingPointTy());
+
+  // add inner_prod call if not already present
+  std::string prod_name =
+      (blas.floatType + "__enzyme_inner_prod" + blas.suffix).str();
+  auto FInnerProdT =
+      FunctionType::get(fpTy, {IT, IT, BlasPT, IT, BlasPT}, false);
+  Function *F =
+      cast<Function>(M.getOrInsertFunction(prod_name, FInnerProdT).getCallee());
+
+  if (!F->empty())
+    return B.CreateCall(F, args, bundles);
+
+  // add dot call if not already present
+  std::string dot_name =
+      (blas.prefix + blas.floatType + "dot" + blas.suffix).str();
+  auto FDotT =
+      FunctionType::get(fpTy, {BlasIT, BlasPT, BlasIT, BlasPT, BlasIT}, false);
+  Function *FDot =
+      cast<Function>(M.getOrInsertFunction(dot_name, FDotT).getCallee());
+
+  // now add the implementation for the inner_prod call
+  F->setLinkage(Function::LinkageTypes::InternalLinkage);
+#if LLVM_VERSION_MAJOR >= 16
+  F->setOnlyAccessesArgMemory();
+#else
+  F->addFnAttr(Attribute::ArgMemOnly);
+#endif
```

Comment:
You might as well mark the whole function as read only as well, no?

Reply:
done, thx